lib/pbench/server/api/resources/server_configuration.py

```diff
@@ -113,40 +113,74 @@ def _put(self, params: ApiParams, _) -> Response:
-                try:
-                    value = params.body["value"]
-                except KeyError:
+        # First, we check for the ways to set an individual config setting by
+        # naming the config key on the URI and specifying a value either with
+        # the "value" either as a query parameter or in the JSON request body.
+        #
+        # We'll complain about JSON request body parameters that are "shadowed"
+        # by the URI or query parameter and might represent client confusion.
+        keydup = set()
+        key = params.uri.get("key")
+        if key:
+            # If we have a key in the URL, the we need a "value" for it, which
+            # we can take either from a query parameter or from the JSON
+            # request payload.
+            if params.body and list(params.body.keys()) != ["value"]:
+                keydup.update(params.body.keys())
+            value = params.query.get("value")
+            if value:
+                if params.body:
+                    keydup.update(params.body.keys())
+            else:
+                value = params.body.get("value")
+                if not value:
                     raise APIAbort(
                         HTTPStatus.BAD_REQUEST,
-                        "Incorrect value specification for key {key!r}",
+                        f"No value found for key system configuration key {key!r}",
                     )
+
+            if len(keydup) > 0:
+                raise APIAbort(
+                    HTTPStatus.BAD_REQUEST,
+                    f"Redundant parameters specified in the JSON request body: {sorted(keydup)!r}",
+                )
+
             try:
                 ServerConfig.set(key=key, value=value)
                 return jsonify({key: value})
             except ServerConfigBadValue as e:
                 raise APIAbort(HTTPStatus.BAD_REQUEST, str(e))
             except ServerConfigError as e:
                 raise APIAbort(HTTPStatus.INTERNAL_SERVER_ERROR, str(e))
-        except KeyError:
-            pass
-
-        try:
-            config = params.body["config"]
-        except KeyError:
-            raise MissingParameters(["config"])
 
-        failures = []
-        for k, v in config.items():
-            try:
-                ServerConfig.set(key=k, value=v)
-            except ServerConfigError as e:
-                self.logger.warning("Unable to update key {} = {!r}: {}", k, v, str(e))
-                failures.append(k)
-        if failures:
-            raise APIAbort(HTTPStatus.INTERNAL_SERVER_ERROR)
-        return jsonify(ServerConfig.get_all())
+        else:
+            # If we get here, we didn't get "key" from the URI, so we're expecting
+            # a JSON request body with key/value pairs, potentially for multiple
+            # config options.
+            badkeys = []
+            for k, v in params.body.items():
+                if k not in ServerConfig.KEYS:
+                    badkeys.append(k)
```

Comment:
You like this better than the list comprehension?
```
            badkeys = [k for k in params.body.keys() if k not in ServerConfig.KEYS]
```
Yeah, Black will probably trash that...c'est la vie.

BTW, I don't see the code using `v` here, so you can use `keys()` instead of `items()`.

Reply:
Yeah, I love comprehensions, in theory, but often find myself avoiding them because of `black`.